Rounded rectangle: add support for distinct X and Y radii

swharden opened the pull request:
This PR extends `ICanvas` to support drawing and filling rounded rectangles with user-defined horizontal and vertical corner radii by creating a new `DrawRoundedRectangle()` and `FillRoundedRectangle()` overloads that accept `xRadius` and `yRadius`. Lack of this functionality was noted by @wieslawsoltes in #268.

To simplify review I did not modify the existing [`PathF.AppendRoundedRectangle()`](https://github.com/dotnet/Microsoft.Maui.Graphics/blob/20e0e9e8c78d37575c7c0e1a5c64170616085f0a/src/Microsoft.Maui.Graphics/PathF.cs#L1098-L1125) but these two methods could be collapsed into one. If you want me to do this in PR or make any other changes, let me know.

I also added test patterns to demonstrate this new functionality:
`DrawRoundedRectangle()` | `FillRoundedRectangle()`
---|---
![image](https://user-images.githubusercontent.com/4165489/148701615-011d5a64-3198-41f1-b06f-a04afaf6d482.png)|![image](https://user-images.githubusercontent.com/4165489/148701628-86f8c78f-b9a0-4bd8-bbe1-fb6b1a99252f.png)

jonlipsky commented:
@swharden Thank you for submitting this PR.  This would be a useful addition to the library; however, I think we need to update the logic slightly before we can accept the PR.  As you may or may not know, you can't reproduce a rounded corner with a cubic bézier curve, though you can do a very close approximation.  As an example, I have attached the following image:

<img width="812" alt="image" src="https://user-images.githubusercontent.com/1466386/150238254-b8551b9f-c2dc-4d51-b9e7-c24285c369c0.png">

First, the shape on the left is a true rounded rectangle (with circles drawn over each corner).   Second, the shape in the middle is a rounded rectangle that has been created with the [existing](https://github.com/dotnet/Microsoft.Maui.Graphics/blob/bcc5019216b3a76952431be6ce0199ee55682a0a/src/Microsoft.Maui.Graphics/PathF.cs#L1101) "AppendRoundedRect" logic present in Maui.Graphics, which is then stretched to twice it's height (with eclipses drawn over each corner).  Third, the shape on the right is drawn using the logic you submitted.  If you look closely at the ellipses that are drawn over each corner, you'll see that there is a visible variation; whereas in the middle version the variation in imperceptable.

The existing code shows how to use the magic ratio of ".55" to calculate the position of the cubic bezier control points to approximate an elliptical curve.  If you make that change, we can accept the PR.

cc: @Redth @mattleibow

swharden commented:
Hi @jonlipsky, thank you for reviewing this PR and providing such an informative response! Your picture was exceptionally helpful.

I was not familiar with this topic but I learned more by reading https://pomax.github.io/bezierinfo/ ... It looks like the maximum deflection between a cubic Bézier curve and a perfect circle can be minimized using a ratio of `0.551784777779014` but there are other other slightly different ratios which minimize different types of errors. I modified my PR to use `.55f` as suggested and consistent with the [existing `AppendRoundedRectangle()`](https://github.com/dotnet/Microsoft.Maui.Graphics/blob/bcc5019216b3a76952431be6ce0199ee55682a0a/src/Microsoft.Maui.Graphics/PathF.cs#L1101), but it may be worth considering using a well-named `const` with full available precision for the dozen instances this number is in the code to help others discover the meaning behind this magic number in case they run across it too. I'm happy to do this in this PR or a new one based on your feedback.

I added a new graphical test to the `DrawRoundedRectangles` scenario confirm this is working as intended.

### Before ❌

rounded rectangle | circularity test | magnified
---|---|---
![image](https://user-images.githubusercontent.com/4165489/150269855-8e14da7d-5ec4-44e0-a7f2-b513f05f52b0.png)|![image](https://user-images.githubusercontent.com/4165489/150269869-8bc9e0e4-aa46-4811-ab13-3513d962a972.png)|![image](https://user-images.githubusercontent.com/4165489/150269894-4fb4219e-8c46-4621-9f5b-477944b7453a.png)

### After ✔️

rounded rectangle | circularity test | magnified
---|---|---
![image](https://user-images.githubusercontent.com/4165489/150269655-9bcd6415-19be-4420-8686-72831e7a3af7.png)|![image](https://user-images.githubusercontent.com/4165489/150269626-3c3e0579-07b7-43d7-ab56-6e4783f836f1.png)|![image](https://user-images.githubusercontent.com/4165489/150269722-9520b1ee-6eb5-4d89-8a1a-b56121234140.png)

jonlipsky commented:
@swharden I like your suggestion of using a single const (with possibly higher precision too) that documents what it's purpose is.  Since you offered, would you mind making that change too, and we'll get these changes accepted into the code base ASAP?

swharden commented:
> would you mind making that change too, and we'll get these changes accepted into the code base ASAP?

Done! 🚀

PS: It would be great if someone ran the autoformatter against the whole code base, or helpful even if this were done one project at a time. I keep almost accidentally committing line changes cuased by the autoformatter on my end...

```
dotnet tool update -g dotnet-format
dotnet-format
```

jonlipsky commented:
@swharden Thanks!  Approved and merged.
